Add a new chapter to define concepts related to pattern-matching.

gafter opened the pull request:
I want to add a new chapter because this will get a lot more complicated
later, for example when we add recursive patterns.

This is proposed to replace #61.

gafter committed:
Add a new chapter to define concepts related to pattern-matching.

I want to add a new chapter because this will get a lot more complicated
later, for exanmple when we add recursive patterns.

gafter committed:
Some minor improvements.

jskeet commented on standard/expressions.md:
```diff
@@ -6124,6 +6137,7 @@ Constant expressions are required in the contexts listed below and this is indic
 - `goto case` statements ([§12.10.4](statements.md#12104-the-goto-statement))
 - Dimension lengths in an array creation expression ([§11.7.15.5](expressions.md#117155-array-creation-expressions)) that includes an initializer.
 - Attributes ([§21](attributes.md#21-attributes))
+- In a *conatant_pattern* (§constant-pattern-new-clause)
```
```suggestion
- In a *constant_pattern* (§constant-pattern-new-clause)
```

MadsTorgersen replied:
```suggestion
- In a *constant_pattern* (§constant-pattern-new-clause)
```

BillWagner commented:
Leaving notes for discussion Wednesday on how to add a chapter and ensure our tools work:

The order of clauses is defined in [clauses.json](https://github.com/dotnet/csharpstandard/blob/draft-v7/standard/clauses.json) for the renumbering tool. The markdown converter relies on the section numbers added by the renumbering tool.

Three other tools all have arrays of filenames in the shell scripts that run the tool:

- [smarten](https://github.com/dotnet/csharpstandard/blob/draft-v7/tools/run-smarten.sh)
- [update grammar annex](https://github.com/dotnet/csharpstandard/blob/draft-v7/tools/update-grammar-annex.sh)
- [validate grammar](https://github.com/dotnet/csharpstandard/blob/draft-v7/tools/validate-grammar.sh)

I can take on the task to update those tools. I'll propose keying all the tools off the `clauses.json` file.

MadsTorgersen commented on standard/expressions.md:
```diff
@@ -3700,6 +3700,7 @@ relational_expression
     | relational_expression '<=' shift_expression
     | relational_expression '>=' shift_expression
     | relational_expression 'is' type
+    | relational_expression 'is' pattern
```
In the future there will be a type pattern, and these two forms of `is` operator will reunify. In anticipation of that I wonder if we should avoid splitting them out quite so deeply here and instead have a single production like

``` antlr
    | relational_expression 'is' pattern_or_type
```

gafter replied:
We need to deal with the fact that the right-hand-side has to have a precedence tighter than expression, or at least as tight as relational_expression.  We may have to bifurcate the grammar for "pattern".

gafter replied:
Also weird lookup: right-hand-side first as a type then as an expression.  This probably ought to be documented, but it isn't urgent and could also be swept under the rug.

gafter replied:
A pattern can derive an expression (via `constant_pattern` and `constant_expression`), but in this context the expression must have precedence of `shift_expression` or tighter (or there would be a grammar ambiguity).  We should open an issue to reflect the fact that work to specify that fact still needs to be done someday.  Alternately, we could just say that in a note in the spec and leave the grammar munging as an exercise to the reader.

BillWagner replied:
Adding the note in #934

BillWagner commented on standard/expressions.md:
```diff
@@ -4033,7 +4034,11 @@ where `x` is an expression of a nullable value type, if operator overload resolu
+
+#### 11.11.11.1 The is-type operator
+
+The *is-type operator* is used to check if the run-time type of an object is compatible with a given type. The check is performed at runtime. The result of the operation `E is T`, where `E` is an expression and `T` is a type other than `dynamic`, is a Boolean value indicating whether `E` is non-null and can successfully be converted to type `T` by a reference conversion, a boxing conversion, an u ...
```
Is "compatible" the right term here? I don't have a better term, but I'm concerned because only the listed conversions are in used.

gafter replied:
This is the existing paragraph.  Such introductory paragraphs are often informal and then clarified by the text that follows.  Where I used that pattern, I introduced the text with "Informally, ".  See a couple of examples in this PR.

gafter replied:
I think the meaning of `compatible` in this context is clear, as it is explained in the rest of the paragraph.  If it isn't clear, we could add  the parenthetical phrase "in the sense defined below" or something of the sort after the word "compatible".  I think this can be left to be dealt with later.

BillWagner replied:
I raised it originally, so I'm happy leaving as is, unless we get feedback outside the committee.

MadsTorgersen commented on standard/expressions.md:
```diff
@@ -4033,7 +4034,11 @@ where `x` is an expression of a nullable value type, if operator overload resolu
-The `is` operator is used to check if the run-time type of an object is compatible with a given type. The check is performed at runtime. The result of the operation `E is T`, where `E` is an expression and `T` is a type other than `dynamic`, is a Boolean value indicating whether `E` is non-null and can successfully be converted to type `T` by a reference conversion, a boxing conversion, an unboxi ...
+There are two forms of the `is` operator.  One is the *is-type operator*, which has a type on the right-hand-side.  The other is the *is-pattern operator*, which has a pattern on the right-hand-side.
+
+#### 11.11.11.1 The is-type operator
```
Could we simply retcon the behavior of the is-type operator as a shorthand for `E is T _`? There seems to be an opportunity for much less duplication here.

gafter replied:
They are different.

https://sharplab.io/#v2:CYLg1APgAgTAjAWAFBQMwAJboMLoN7LpGYZQAs6AsgBQCU+hxTAlgGbrVzrMDO6AcgFcANsICGAI2EBTADzMAdgBcAfPTwBfRkyJsOXXgJHipcxavQB9dVqRNbt5FgCC+W0A

MadsTorgersen commented on standard/lexical-structure.md:
```diff
@@ -65,13 +65,14 @@ The productions for *simple_name* ([§11.7.4](expressions.md#1174-simple-names))
+- In certain contexts, we treat *identifier* as a disambiguating token. Those contexts are where the sequence of tokens being disambiguated is immediately preceded by one of the keywords `is`, `case` or `out`, or arises while parsing the first element of a tuple literal (in which case the tokens are preceded by `(` or `:` and the identifier is followed by a `,`) or a subsequent element of a tuple ...
 
-then the *type_argument_list* is retained as part of the *simple_name*, *member_access*, or *pointer_member_access* and any other possible parse of the sequence of tokens is discarded. Otherwise, the *type_argument_list* is not considered part of the *simple_name*, *member_access*, or *pointer_member_access*, even if there is no other possible parse of the sequence of tokens.
+If the following token is among this list, or an identifier in such a context, then the *type_argument_list* is retained as part of the *simple_name*, *member_access* or  *pointer_member-access* and any other possible parse of the sequence of tokens is discarded.  Otherwise, the *type_argument_list* is not considered to be part of the *simple_name*, *member_access* or *pointer_member_access*, eve ...
```
I have a bit of a hard time following the flow of this rule, as it first "sees if" certain conditions hold, and then later draws conclusions from that. Probably a bit of wordsmithing is all it takes.

gafter replied:
I agree.  I don't think there is any urgency to address this immediately.

BillWagner replied:
Logged #935 to address later.

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+I would suggest placing this new chapter between the existing 12 (Statements) and 13 (namespaces)
```
Another placement would be before the existing 11 (Expressions). We tend to introduce building blocks before we use them, and patterns are used by expressions and statements. 

Another option is to not make them a top level chapter but add them to the set of sections in the beginning of Expressions providing building blocks for the subsequent syntax-based sections. This is what I have proposed to do with "Deconstruction" in #644, and pattern matching in a later version of C# will make use of deconstruction, so having patterns follow deconstruction would fit with that.

gafter replied:
Before expressions makes sense to me.  I suggested after only because expressions seem like a more fundamental and important concept, and each uses the other.

The patterns "chapter" will get more complicated when we add more pattern concepts in the next couple of versions (and in the future if discriminated unions are added), so I wanted to break them out so as not to interrupt the flow of the expressions description.

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+
+## §patterns-new-clause-general General
+
+A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested in the context of a switch expression or a *relational_expression* that is on the left-hand s ...
```
```suggestion
A ***pattern*** is a syntactic form that can be used with the `is` operator ([§11.11.11](expressions.md#111111-the-is-operator)) and in a *switch_statement* ([§12.8.3](statements.md#1283-the-switch-statement)) to express the shape of data against which incoming data is to be compared. A pattern is tested against the *expression* of a switch statement, or against a *relational_expression* that is on the left-hand side of an `is` operator. We call this a ***pattern input value***.
```

BillWagner commented on standard/statements.md:
````diff
@@ -640,29 +640,54 @@ switch_section
 ```
 
-A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§12.3.2](statements.md#1232-statement-lists)).
+A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§12.3.2](statements.md#1232-statement-lists)). Each *switch ...
````
nit, missing word:

```suggestion
A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§12.3.2](statements.md#1232-statement-lists)). Each *switch_label* containing `case` has an associated pattern (§patterns-new-clause) against which the value of the switch expression is tested (XREF NEEDED). If *case-guard* is present, its expression shall be implicitly convertible to the type `bool` and that expression is evaluated as an additional condition to be satisfied for the case to be considered satisfied.
```

MadsTorgersen replied:
```suggestion
A *switch_statement* consists of the keyword `switch`, followed by a parenthesized expression (called the ***switch expression***), followed by a *switch_block*. The *switch_block* consists of zero or more *switch_section*s, enclosed in braces. Each *switch_section* consists of one or more *switch_label*s followed by a *statement_list* ([§12.3.2](statements.md#1232-statement-lists)). Each *switch_label* containing `case` has an associated pattern (§patterns-new-clause) against which the value of the switch expression is tested (XREF NEEDED). If *case-guard* is present, its expression shall be implicitly convertible to the type `bool` and that expression is evaluated as an additional condition for the case to be considered satisfied.
```

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.
+
+Each pattern form defines the set of types for input values that that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
```
```suggestion
Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
```

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+
+### §declaration-pattern-new-clause Declaration pattern
+
+A *declaration_pattern* is used to test that a value has a given type and if the test succeeds, to cast that value to that type.
```
The term "cast" seems potentially suggestive of different behavior.

```suggestion
A *declaration_pattern* is used to test that a value has a given type and, if the test succeeds, provide the value in a variable of that type.
```

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+declaration_pattern
+    : type simple_designation
+    ;
+simple_designation
```
Remind me why we have this chain of trivial productions - is it in anticipation of future additions?

gafter replied:
Yes, we have tuple-like designations in the future, and it becomes optional.

MadsTorgersen commented on standard/patterns.md:
````diff
@@ -0,0 +1,172 @@
+    ;
+```
+
+The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.
````
> If it is of that runtime type (or some subtype), the pattern *matches* that value.

This seems underspecified. Compare to the detail in the specification of the existing type-based `is` operator. In fact I'd suggest moving all that detail here and respecifying the `E is T` operator as simply a shorthand for `E is T _`.

gafter replied:
The is-type operator permits a nullable type on the right-hand-side.  The is-pattern operator does not.  But we can still unify most of it.

Nigel-Ecma replied:
Also should "subtype" be "pattern-compatilble type"?

BillWagner replied:
Addressed in #934 

reworded to use the rules in `is type`, with the added rule that the type can't be a nullable value type.

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+
+The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.
+
+Given a *pattern input value* *e*, if the *simple_designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause) the value of *e* is not bound to anything. (Although a declared variable with the name `_` may be in scope at that point, that named variable is not seen in this context.) If *simple_designation* is any other identifier, a local variable ([§9.2.8](variables.md#928-lo ...
```
This may need a bit of tinkering once this PR meets up with the discards coming in from #664.

gafter replied:
Maybe, maybe not.  I suggest creating an issue to examine this in detail at some later time.  I suspect no action is needed.

BillWagner replied:
Logged in #936

MadsTorgersen commented on standard/patterns.md:
````diff
@@ -0,0 +1,172 @@
+> ```
+>
+> *end example*
+It is an error if *type* is a nullable value type.
````
Missing blank line after `> *end example*`

gafter replied:
```suggestion

It is an error if *type* is a nullable value type.
```

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+> *end example*
+It is an error if *type* is a nullable value type.
+
+> *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and the type of `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment
```
```suggestion
> *Example*: The declaration pattern can be used to test values of nullable types: a value of type `Nullable<T>` (or a boxed `T`) matches a type pattern `T2 id` if the value is non-null and `T2` is `T`, or some base type or interface of `T`. For example, in the code fragment
```

BillWagner commented on standard/variables.md:
```diff
@@ -287,8 +287,49 @@ if ( «expr» ) «then_stmt» else «else_stmt»
+The definite-assignment state of *v* on the control flow transfer to a reachable switch block statement list is
+- If the control transfer was due to a 'goto case' or 'goto default' statement, then the state of *v* is the same as the state at the beginning of that 'goto' statement.
+- If the control transfer was due to the `default` label of the switch, then the state of *v* is the same as the state of *v* after *expr*.
+- If the control transfer was due to an unreachable switch label, then the state of *v* is "definitely assigned".
```
This one confuses me. I think I need an example to understand the implications of v being definitely assigned due to an unreachable switch label.

gafter replied:
```
bool b;
switch (1) {
  case 2 when b: // b is definitely assigned here.
    break;
}
```

gafter replied:
I don't think this is a common enough case to deserve an example in the text, though I admit it would help the reader make sense of the text.  This can be left to be dealt with later.

BillWagner replied:
Logged #937 to discuss later.

BillWagner commented on standard/patterns.md:
````diff
@@ -0,0 +1,172 @@
+    ;
+```
+
+The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.
````
This should include boxing and unboxing conversions, right?

gafter replied:
Yes.  We should unify with the description of the is-type operator.

gafter replied:
This section should refer to the specification of the is-type operator for the meaning of when it is true or false.  This is probably better done sooner rather than later.

BillWagner replied:
Done in #934

MadsTorgersen commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+Here we define when a set of patterns is *exhaustive* for a type.
+
+A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:
+1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would that value; or
```
```suggestion
1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would match that value; or
```

BillWagner reviewed (commented):
I do like the idea of a new clause for patterns. I had a few comments and nits, but otherwise this is great.

gafter commented on standard/expressions.md:
```diff
@@ -4033,7 +4034,11 @@ where `x` is an expression of a nullable value type, if operator overload resolu
+
+The *is-type operator* is used to check if the run-time type of an object is compatible with a given type. The check is performed at runtime. The result of the operation `E is T`, where `E` is an expression and `T` is a type other than `dynamic`, is a Boolean value indicating whether `E` is non-null and can successfully be converted to type `T` by a reference conversion, a boxing conversion, an u ...
 
 The operation is evaluated as follows:
```
E is no longer permitted to be a method group.  I don't know when that change happened.

gafter replied:
Also "of if" below should be "or if".  Better sooner rather than later.

gafter replied:
Two changes should be made sooner rather than later:
1. We should document the error `error CS0837: The first operand of an 'is' or 'as' operator may not be a lambda expression, anonymous method, or method group.` and update this section accordingly.
2. Change the words "of if" below to "or if".

BillWagner replied:
Added to #934

gafter commented on standard/expressions.md:
```diff
@@ -4071,6 +4076,14 @@ User defined conversions are not considered by the `is` operator.
+The *is-pattern operator* is used to check if the value computed by an expression *matches* a given pattern (XREF TO DEF OF "PATTERN MATCHES"). The check is performed at runtime. The result of the is-pattern operator is true if the value matches the pattern; otherwise it is false.
+
+For an expression of the form `E is P`, where `E` is a relational expression of type `T` and `P` is a pattern, it is a compile-time error if any of the following hold:
+- `E` does not designate a value or does not have a type.
```
```suggestion

- `E` does not designate a value or does not have a type.
```

gafter commented on standard/patterns.md:
````diff
@@ -0,0 +1,172 @@
+> if (expr is Type v) { /* code using v */ }
+> ```
+>
+> *end example*
````
```suggestion
> *end example*

```

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+A constant pattern `P` is *applicable to* a type `T` if there is an implicit conversion from the constant expression of `P` to the type `T`.
+
+For a constant pattern `P`, we say its *converted value* is
+- if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise
```
```suggestion

- if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise
```

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+- the value of the pattern's constant value.
+
+Given a *pattern input value* *e* and a a constant pattern `P` with converted value *v*,
+- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
```
```suggestion

- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
```

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+We say a pattern `P` *would match* a constant `K` if the specification for that pattern's runtime behavior is that `P` matches `K`.
+
+A set of patterns `Q` *subsumes* a pattern `P` if any of the following conditions hold:
+- `P` is a constant pattern and any of the patterns in the set `Q` would match `P`'s *converted value*
```
```suggestion

- `P` is a constant pattern and any of the patterns in the set `Q` would match `P`'s *converted value*
```

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,172 @@
+Informally, we say that a set of patterns is exhaustive for a type if some pattern in the set is applicable to every possible value of that type other than null.
+Here we define when a set of patterns is *exhaustive* for a type.
+
+A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:
```
```suggestion
A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:

```

gafter commented on standard/statements.md:
```diff
@@ -830,18 +855,45 @@ When the governing type of a `switch` statement is `string` or a nullable value
-- The switch expression is a constant value that doesn’t match any `case` label, and the switch section contains the `default` label.
-- A switch label of the switch section is referenced by a reachable `goto case` or `goto default` statement.
+A switch label is reachable if at least one of the following is true:
+- The switch expression is a constant value and either
```
```suggestion

- The switch expression is a constant value and either
```

gafter commented on standard/statements.md:
```diff
@@ -830,18 +855,45 @@ When the governing type of a `switch` statement is `string` or a nullable value
-The end point of a `switch` statement is reachable if at least one of the following is true:
+The statement list of a given switch section is reachable if the `switch` statement is reachable and the switch section contains a reachable switch label.
 
+The end point of a `switch` statement is reachable if the switch statement is reachable and at least one of the following is true:
```
```suggestion
The end point of a `switch` statement is reachable if the switch statement is reachable and at least one of the following is true:

```

gafter committed:
Apply suggestions from code review

Co-authored-by: Mads Torgersen <mads.torgersen@microsoft.com>

jskeet commented on standard/patterns.md:
````diff
@@ -0,0 +1,176 @@
+    ;
+```
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.
````
*Can* result, or *always* result (but possibly without the variable being definitely assigned)?

jskeet replied:
Ah - discards, right?

gafter replied:
Right.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+A *declaration_pattern* and a *var_pattern* can result in the declaration of a local variable.
+
+Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
```
I think we may want an example here to clarify this. I believe I get it, but it could be confusing.

For example, consider:

```csharp
if (x is string s)
```

It's fine for the compile-time type of `x` to be `object`, even though a value whose execution type is `object` *won't* match. But the expression type being `object` is okay because the value *might* be a `string` reference. Whereas if the compile-time type of `x` is (say) `Stream` then this would be a compile-time error.

I think that sort of clarification can be added later - I don't want us to wait for perfection before merging this, so long as the text is accurate when read appropriately (and I think it is here). We might want to keep a list of this sort of thing somewhere, so we can come back afterwards. (It's inevitable when adding a huge chunk this PR does.)

gafter replied:
Mention this is a compile-time concept.

The simplest way to do that is change "it is an error" to "it is a compile-time error".

gafter replied:
Agreed, that expanding upon this, perhaps with a sentence or two just as you wrote here, is worth doing (later)

BillWagner replied:
"compile time" added in #934 

#938 logs the request for an example.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+The runtime type of the value is tested against the *type* in the pattern. If it is of that runtime type (or some subtype), the pattern *matches* that value. This pattern form never matches a `null` value.
+
+Given a *pattern input value* *e*, if the *simple_designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause) the value of *e* is not bound to anything. (Although a declared variable with the name `_` may be in scope at that point, that named variable is not seen in this context.) If *simple_designation* is any other identifier, a local variable ([§9.2.8](variables.md#928-lo ...
```
I suspect the rendering of *pattern input value* *e* would be clearer without "pattern input value" in italics. Or we could try to reword, something like "Given a *pattern input value* of the form *e*" - although that implies there are other forms. Thoughts welcome.

gafter replied:
I'm open to seeing the wording reconsidered.  Also, the type matching is more than just subtype.  I believe it should delegate the specification to the `is-type` operator (e.g. include boxing and unboxing).  Not urgent.

BillWagner replied:
Already addressed in earlier edits in #934

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+Given a *pattern input value* *e*, if the *simple_designation* is the *identifier* `_`, it denotes a discard (§discards-new-clause) the value of *e* is not bound to anything. (Although a declared variable with the name `_` may be in scope at that point, that named variable is not seen in this context.) If *simple_designation* is any other identifier, a local variable ([§9.2.8](variables.md#928-lo ...
+
+Certain combinations of static type of the pattern input value and the given type are considered incompatible and result in a compile-time error. A value of static type `E` is said to be ***pattern compatible*** with the type `T` if there exists an identity conversion, an implicit reference conversion, a boxing conversion, an explicit reference conversion, or an unboxing conversion from `E` to `T ...
```
Can we reduce "an implicit reference conversion, a boxing conversion, an explicit reference conversion" to "a reference conversion, a boxing conversion, ..."?

jskeet replied:
Nit: the "period double-space" before "A declaration pattern" makes me twitch. I'm happy to change those afterwards :)

gafter replied:
> Can we reduce "an implicit reference conversion, a boxing conversion, an explicit reference conversion" to "a reference conversion, a boxing conversion, ..."?

We don't define the phrase "reference conversion".  However, we could say "implicit or explicit reference conversion".  However, this is super low priority.

> Nit: the "period double-space" before "A declaration pattern" makes me twitch. I'm happy to change those afterwards :)

One of us will end up twitching.

BillWagner replied:
I added the change to #934 (because it was even easier than logging the low priority issue).

I checked and removed the remaining instances of "period double-space" exist in any clause, except where the two spaces was used at the end of a line to force a line break.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+- if the input expression's type is the nullable version of an integral type or an enum type, the pattern's constant value converted to its underlying type; otherwise
+- the value of the pattern's constant value.
+
+Given a *pattern input value* *e* and a a constant pattern `P` with converted value *v*,
```
```suggestion
Given a *pattern input value* *e* and a constant pattern `P` with converted value *v*,
```

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
+- the pattern `P` *matches* the value *e* if `object.Equals(e, v)` returns `true`.
+
+> *Example*:
```
I think this example could do with a bit more explanation - particularly as unless you're attuned to it already, you might be wondering where the pattern is.

(Also, please put the opening brace on a new line.)

gafter replied:
Please propose an edit.

gafter replied:
This minor typographical change can be deferred.

BillWagner replied:
Added a sentence, and moved the opening brace.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+Given a *pattern input value* *e* and a a constant pattern `P` with converted value *v*,
+
+- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
+- the pattern `P` *matches* the value *e* if `object.Equals(e, v)` returns `true`.
```
It looks like the compiler is allowed to generate code with the same effect as that, but without actually calling `object.Equals` (which will box for float/decimal/double of course). For example, I've just tried `d is double.NaN` and Roslyn compiles that to `double.IsNaN(d)`.
Do we need to mention that?

gafter replied:
Generally, no.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+## Pattern Exhaustiveness
+
+Informally, we say that a set of patterns is exhaustive for a type if some pattern in the set is applicable to every possible value of that type other than null.
```
There are two ways of reading "if some pattern in the set is applicable to every possible value of that type" - my initial reading is "one of the patterns in the set has the property that it matches every possible value of that type". Whereas I assume you really mean "if every possible value of that type is matched by at least one pattern in the set".

Let's add this to the "non-blocking, could do with tweaking" list - assuming I've read it correctly.

gafter replied:
Agreed.

BillWagner replied:
Logged in #939

jskeet commented on standard/statements.md:
```diff
@@ -640,29 +640,54 @@ switch_section
+>     case var _: // error: pattern subsumed, as previous case always matches
+>         break;
+>     default:
+>         break;  // warning: unreachable,
```
Incomplete comment?

gafter replied:
Just remove the comma.  Or add an explanation "all possible values already handled".

Low priority.

BillWagner replied:
Made the edit in #934. More efficient than logging a new issue.

jskeet commented on standard/statements.md:
```diff
@@ -830,18 +855,47 @@ When the governing type of a `switch` statement is `string` or a nullable value
+- The switch expression is not a constant value and either
+  - the label is a `case` without a guard or with a guard whose value is not the constant false; or
+  - it is a `default` label and
+    - the set of patterns appearing among the cases of the switch statement that do not have guards or have guards whose value is the constant true, is not *exhaustive* (NEED XREF) for the switch controlling type; or
```
governing type here and below? (Instead of controlling type.)

gafter replied:
Yes, please change.

BillWagner replied:
Added to #934

jskeet commented on standard/statements.md:
```diff
@@ -830,18 +855,47 @@ When the governing type of a `switch` statement is `string` or a nullable value
+    - The switch expression is a non-constant value of a nullable type, and no pattern appearing among the cases of the switch statement that do not have guards or have guards whose value is the constant true would match the value `null` (XREF to "would match" in patterns).
+    - The switch expression is a constant value and no `case` label without a guard or whose guard is the constant true would match that value.
+
+> *Example*: The following code shows a succinct use of the when clause:
```
```suggestion
> *Example*: The following code shows a succinct use of the `when` clause:
```

gafter replied:
Yes, should change.  Low priority, but also trivial to do.

BillWagner replied:
added to #934

jskeet commented on standard/statements.md:
```diff
@@ -830,18 +855,47 @@ When the governing type of a `switch` statement is `string` or a nullable value
+>         case "circle":
+>             return new Circle(2);
+>         …
+>         case var o when (o?.Trim().Length ?? 0) == 0:
```
I'd normally write this as `case var o when string.IsNullOrEmpty(o?.Trim())`

Happy to discuss merits either way (with the context of this example).

jskeet replied:
(I suppose actually I'd use `string.IsNullWhiteSpace` actually...)

gafter replied:
It's not wrong, but I agree that would be better guidance.  Low priority.

BillWagner replied:
trivial, so I made the edit.

jskeet reviewed (commented):
Various comments, but I like having patterns in their own section. Let's talk in the meeting about concrete steps going from here to merged.

(This is the sort of thing where I'm a terribly judge of whether it's actually "done" or not - there are any number of paragraphs that Neal could have omitted and I'd never have noticed...)

Nigel-Ecma commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+### §declaration-pattern-new-clause Declaration pattern
+
+A *declaration_pattern* is used to test that a value has a given type and, if the test succeeds, provide the value in a variable of that type.
```
From below its "a value has a type which is pattern compatible with a given type" – which is a bit long winded so some wordsmithing is required…

gafter replied:
No, this is a different, runtime concept.

(no action needed)

gafter replied:
Should take its description from the is-type operator.

Best addressed sooner rather than later.

gafter replied:
This would be better done sooner rather than later.  Two things:
1. The specification about the value being of the given type is actually a bit more complex and should delegate to the spec of the is-type operator.  This is a runtime type test.
2. The *pattern compatible* concept is to give warnings based on compile-time types that are impossible to match.  If that isn't clear, it should be made clear.

BillWagner replied:
Addressed in #934

Added text that the runtime test must produced the same results as the is-type operator. Earlier edits should have already clarified the *pattern compatible* concept as a compile-time operation.

jskeet commented on standard/expressions.md:
```diff
@@ -3700,6 +3700,7 @@ relational_expression
     | relational_expression '<=' shift_expression
     | relational_expression '>=' shift_expression
     | relational_expression 'is' type
+    | relational_expression 'is' pattern
```
Meeting notes: there's ambiguity here. We should somehow reference how it's resolved. We *may* be able to merge these, but we'll need to talk about how nullable value types are okay for `type` but not `pattern`.

gafter replied:
Agreed, and also mention that the precedence of an expression, if the pattern is a constant pattern, is restricted to `shift_expression` or tighter.

gafter replied:
Though honestly it could be done later too.

BillWagner replied:
Added to #935

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+Each pattern form defines the set of types for input values that the pattern may be applied to.  We say a pattern `P` is *applicable to* a type `T` if `T` is among the types whose values the pattern may match.  It is an error if a pattern `P` appears in a program to match a *pattern input value* of type `T` if `P` is not applicable to `T`.
+
+Each pattern form defines the set of values for which the pattern *matches* the value.
```
Say this is a runtime concept.  (Not urgent, or even necessary)

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+>
+> *end example*
+
+It is an error if *type* is a nullable value type.
```
... unless it is the type of an is-type expression?

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+For a constant pattern `P`, we say its *converted value* is
+
+- if the input expression's type is an integral type or an enum type, the pattern's constant value converted to that type; otherwise
```
not called input expression...

gafter replied:
Change "input expression" to "pattern input value".

BillWagner replied:
Edited. Check for any additional occurrence of "input expression" where "pattern input value" should be used. None found.

gafter committed:
Update standard/patterns.md

Co-authored-by: Jon Skeet <jonskeet@google.com>

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+
+### §var-pattern-new-clause Var pattern
+
+A *var_pattern* matches every value.  That is, a pattern-matching operation with a *var_pattern* always succeeds.
```
Matches in italics

gafter replied:
Sooner, please.

BillWagner replied:
Added in #934

jskeet commented:
- Bill to add a commit to fix the GH Actions
- Neal can address any trivial issues
- If we can get the new xrefs working before merging, let's try to get that in
- We can then merge, and then fix any remaining nits afterwards
- If we didn't get xrefs working before, do it now.

gafter commented on standard/patterns.md:
```diff
@@ -0,0 +1,176 @@
+Here we define when a set of patterns is *exhaustive* for a type.
+
+A set of patterns `Q` is *exhaustive* for a type `T` if any of the following conditions hold:
+1. `T` is an integral or enum type, or a nullable version of one of those, and for every possible value of `T`'s underlying type, some pattern in `Q` would match that value; or
```
"underlying" is meant to apply to nullable types, not enum types.

gafter replied:
change "underlying" to "nullable underlying".

BillWagner replied:
added in #934

BillWagner committed:
add new clause to chapter config

This fixes the section renumbering tool for the added new clause.

BillWagner commented:
My last commit gets the section renumber tool to work correctly (once #664 is merged).

I'll make a fresh PR to get the other tools using the same config file for chapter sources

BillWagner committed:
update shell scripts

Add the new patterns clause to both the patterns generator and the validate grammar scripts.

BillWagner committed:
remove note before H1

Because the first line isn't an H1, the grammar validator is skipping this file.

BillWagner committed:
Remove old nupkg

Testing the regex for not skipping the new clause.

BillWagner committed:
Add updated nupkg

BillWagner committed:
update version.

jskeet commented on standard/patterns.md:
```diff
@@ -0,0 +1,174 @@
+- if *e* has integral type or enum type, or a nullable form of one of those, and *v* has integral type, the pattern `P` *matches* the value *e* if result of the expression `e == v` is `true`; otherwise
+- the pattern `P` *matches* the value *e* if `object.Equals(e, v)` returns `true`.
+
+> *Example*:
```
```suggestion
> *Example*: The `switch` statement in the following method uses five constant patterns in its case labels.
```

gafter replied:
Should be done sooner rather than later.

jskeet commented:
I'm going to rebase this, fix the conflicts, and merge it, so that we've got it in the Word document draft. The remaining comments will still need to be addressed, but that can happen afterwards.

jskeet committed:
Fix Markdown lint violations

jskeet committed:
Fix final lint issue

BillWagner commented:
> I'm going to rebase this, fix the conflicts, and merge it, so that we've got it in the Word document draft. The remaining comments will still need to be addressed, but that can happen afterwards.

Good plan. Can you ping me when you merge it? (@ - mention is fine). I'll likely have some downstream links to fix once it's merged, and the sections have been renumbered.

jskeet commented:
@BillWagner: I've temporarily disabled the check for all conversations to be resolved (as we want to address those afterwards) but we need to get the renumbering working. Are you happy to pull this branch, fix that up and repush? I'm hoping the Word converter will work *after* the renumbering is done.

BillWagner commented:
> @BillWagner: I've temporarily disabled the check for all conversations to be resolved (as we want to address those afterwards) but we need to get the renumbering working. Are you happy to pull this branch, fix that up and repush? I'm hoping the Word converter will work _after_ the renumbering is done.

Yes. I can do it. My comment https://github.com/dotnet/csharpstandard/pull/757#issuecomment-1476571863 has my notes from earlier.

BillWagner committed:
fix section numbers.

BillWagner committed:
found it!
